### PR TITLE
[CLOUD_CONTAINER_DELETION] Core logic for container compaction

### PR DESCRIPTION
**Pull request by ankagrawal**
Container deletion logic in cloud.

### CODE REVIEW

File: ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java [hunk 1]
```diff
@@ -123,19 +126,37 @@ public void onPartitionRemoved(PartitionId partitionId) {
...
-          cloudConfig.cloudBlobCompactionIntervalHours, delaySec);
+    // Schedule thread to purge dead blobs for this VCR's partitions
+    // after delay to allow startup to finish.
+    scheduleTask(cloudStorageCompactor, cloudConfig.cloudBlobCompactionEnabled,
+        cloudConfig.cloudBlobCompactionStartupDelaySecs,
+        TimeUnit.HOURS.toSeconds(cloudConfig.cloudBlobCompactionIntervalHours), "cloud blob compaction");
+
+    // Schedule thread to purge blobs belonging to deprecated containers for this VCR's partitions
+    // after delay to allow startup to finish.
+    scheduleTask(() -> cloudContainerCompactor.compactAssignedDeprecatedContainers(
```
Comment (jsjtzyy) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java:140]: `cloudContainerCompactor.compactAssignedDeprecatedContainers` will throw NullPointerException if cloud container compaction is disabled (see line 81)

File: ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java [hunk 2]
```diff
@@ -123,19 +126,37 @@ public void onPartitionRemoved(PartitionId partitionId) {
...
+        virtualReplicatorCluster.getAssignedPartitionIds()), cloudConfig.cloudContainerCompactionEnabled,
+        cloudConfig.cloudContainerCompactionStartupDelaySecs,
+        TimeUnit.HOURS.toSeconds(cloudConfig.cloudContainerCompactionIntervalHours), "cloud container compaction");
+  }
```
Comment (jsjtzyy) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java:144]: According to your design, there would be several threads concurrently performing container compaction. Could you point me to the implementation?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudConfig.java [hunk 3]
```diff
@@ -135,5 +156,11 @@ public AzureCloudConfig(VerifiableProperties verifiableProperties) {
     azureBlobContainerStrategy =
         verifiableProperties.getString(AZURE_BLOB_CONTAINER_STRATEGY, DEFAULT_CONTAINER_STRATEGY);
     azureNameSchemeVersion = verifiableProperties.getInt(AZURE_NAME_SCHEME_VERSION, DEFAULT_NAME_SCHEME_VERSION);
+    cosmosContainerDeletionBatchSize =
+        verifiableProperties.getInt(COSMOS_CONTAINER_DELETION_BATCH_SIZE, DEFAULT_COSMOS_CONTAINER_DELETION_BATCH_SIZE);
+    containerCompactionAbsPurgeLimit =
+        verifiableProperties.getInt(CONTAINER_COMPACTION_ABS_PURGE_LIMIT, DEFAULT_CONTAINER_COMPACTION_ABS_PURGE_LIMIT);
+    containerCompactionCosmosQueryLimit = verifiableProperties.getInt(CONTAINER_COMPACTION_COSMOS_QUERY_LIMIT,
+        DEFAULT_COSMOS_CONTAINER_DELETION_BATCH_SIZE);
```
Comment (jsjtzyy) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudConfig.java:164]: same here, better to use `getIntInRange`

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java [hunk 4]
```diff
@@ -489,10 +493,16 @@ private CloudStorageException toCloudStorageException(String message, Exception
 
   @Override
   public void deprecateContainers(Collection<Container> deletedContainers) throws CloudStorageException {
+    //TODO need to set correct partition class for video cluster in call to getAllPartitionIds.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java:496]: Remove or reword this TODO not to use LI things.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageCompactor.java [hunk 5]
```diff
@@ -264,44 +266,6 @@ private int compactPartitionBucketed(String partitionPath, String fieldName, lon
     }
   }
 
-  /**
-   * Permanently delete the specified blobs in Azure storage.
-   * @param blobMetadataList the list of {@link CloudBlobMetadata} referencing the blobs to purge.
-   * @return the number of blobs successfully purged.
-   * @throws CloudStorageException if the purge operation fails for any blob.
-   */
-  int purgeBlobs(List<CloudBlobMetadata> blobMetadataList) throws CloudStorageException {
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageCompactor.java:273]: Can we keep this method and make it a wrapper around the utility?  Then callers don't need to know the impl details.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java [hunk 6]
```diff
@@ -70,11 +72,12 @@
   private static final String LIMIT_PARAM = "@limit";
   private static final String EXPIRED_BLOBS_QUERY = constructDeadBlobsQuery(CloudBlobMetadata.FIELD_EXPIRATION_TIME);
   private static final String DELETED_BLOBS_QUERY = constructDeadBlobsQuery(CloudBlobMetadata.FIELD_DELETION_TIME);
+  private static final String CONTAINER_BLOBS_QUERY =
+      "SELECT TOP %d * FROM c WHERE c.accountId=%d and c.containerId=%d";
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:76]: Can we use parameters in the new queries (like LIMIT_PARAM) to be consistent?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java [hunk 7]
```diff
@@ -425,6 +428,51 @@ CloudBlobMetadata getMetadataOrNull(BlobId blobId) throws DocumentClientExceptio
...
+   * @param containerId container id of the container.
+   * @param queryLimit max number of blobs to return.
+   * @return a List of {@link CloudBlobMetadata} referencing the blobs belonging to the deprecated containers.
+   * @throws DocumentClientException in case of any error.
+   */
+  List<CloudBlobMetadata> getContainerBlobs(String partitionPath, short accountId, short containerId, int queryLimit)
+      throws DocumentClientException {
+    String query = String.format(CONTAINER_BLOBS_QUERY, accountId, containerId);
+    SqlQuerySpec querySpec = new SqlQuerySpec(query);
+    FeedOptions feedOptions = new FeedOptions();
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:444]: The code from here down is nearly identical to getDeadBlobs.  Please refactor into a common class that take SqlQuerySpec as argument.

Comment (lightningrob) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:444]: Also, since we aren't time bucketing the queries like dead blob compaction does, we could run into the same throttling issues that prompted the use of time bucketing there.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java [hunk 8]
```diff
@@ -559,25 +608,89 @@ public long deprecateContainers(Set<ContainerDeletionEntry> deprecatedContainers
...
+   * Returns {@Null} if the field already has the specified value.
+   * @throws DocumentClientException if the record was not found or if the operation failed.
+   */
+  ResourceResponse<Document> updateContainerDeletionEntry(short containerId, short accountId,
+      BiConsumer<Document, AtomicBoolean> updateFields) throws DocumentClientException {
+
+    // Read the existing record
+    String id = CosmosContainerDeletionEntry.generateContainerDeletionEntryId(accountId, containerId);
+    String docLink = getContainerDeletionEntryDocumentLink(id);
+    RequestOptions options = getRequestOptions(id);
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:657]: A lot of this logic is very similar to updateMetadata().  Possible to reuse/combine?

File: ambry-cloud/src/test/resources/azure-test.properties [hunk 9]
```diff
@@ -9,9 +9,8 @@
 # under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 # CONDITIONS OF ANY KIND, either express or implied.
 #
-azure.storage.connection.string=<storage-account-credentials>
-cosmos.endpoint=<cosmos-url>
-cosmos.collection.link=/dbs/ambry-metadata/colls/blob-metadata
-cosmos.key=<cosmos-key>
-cosmos.deleted.container.collection.link=/dbs/ambry-metadata-main/colls/deleted-container-test
-cosmos.direct.https=true
+azure.storage.connection.string=DefaultEndpointsProtocol=https;AccountName=wus2ambryblobstore1;AccountKey=vqQbP8s6I68IGlyYeqyv1dAwsh3AXXlD+QwgGTVhbU5NEMzE6Tucv9SxS2/6AzB/8VDHjaZmIYV9MWUCWaAqTg==;EndpointSuffix=core.windows.net
```
Comment (lightningrob) [inline ambry-cloud/src/test/resources/azure-test.properties:12]: Ouch!  Obviously remove this.  We should also change the EI keys since they've been exposed.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCompactionUtil.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ */
+package com.github.ambry.cloud.azure;
+
+import com.github.ambry.cloud.CloudBlobMetadata;
+import com.github.ambry.cloud.CloudStorageException;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+
+public class AzureCompactionUtil {
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCompactionUtil.java:22]: Javadoc.

**Review by lightningrob — CHANGES_REQUESTED** [review]
Mostly looks good.  A few concerns mentioned.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosContainerDeletionEntry.java [hunk 11]
```diff
@@ -25,34 +25,36 @@
...
+public class CosmosContainerDeletionEntry {
   static final String VERSION_KEY = "version";
   static final String CONTAINER_ID_KEY = "containerId";
   static final String ACCOUNT_ID_KEY = "accountId";
-  static final String CONTAINER_DELETE_TRIGGER_TIME_KEY = "deleteTriggerTime";
-  static final String IS_DELETED_KEY = "isDeleted";
+  static final String CONTAINER_DELETE_TRIGGER_TIME_KEY = "deleteTriggerTimestamp";
+  static final String DELETED_KEY = "deleted";
   static final String DELETE_PENDING_PARTITIONS_KEY = "deletePendingPartitions";
+  private static final String ID_KEY = "id";
```
Comment (SophieGuo410) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosContainerDeletionEntry.java:35]: Not used?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureContainerCompactor.java [hunk 12]
```diff
@@ -53,38 +66,83 @@
...
+  public void compactAssignedDeprecatedContainers(List<? extends PartitionId> assignedPartitions) {
+    try {
+      SortedSet<CosmosContainerDeletionEntry> containerDeletionEntrySet =
+          fetchContainerDeletionEntries(assignedPartitions);
+      while (!containerDeletionEntrySet.isEmpty()) {
+        CosmosContainerDeletionEntry containerDeletionEntry = containerDeletionEntrySet.first();
+        containerDeletionEntrySet.remove(containerDeletionEntry);
+        for (String partitionId : containerDeletionEntry.getDeletePendingPartitions()) {
+          try {
+            int blobCompactedCount =
```
Comment (SophieGuo410) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureContainerCompactor.java:121]: Seems the blobCompactedCount never been used?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureContainerCompactor.java [hunk 13]
```diff
@@ -131,4 +193,96 @@ private void saveLatestContainerDeletionTime(long latestContainerDeletionTimesta
...
+
+  /**
+   * Fetch the {@link CosmosContainerDeletionEntry} from cloud and create a cache with entries that have atleast one partition
+   * assigned to current node.
+   */
+  private SortedSet<CosmosContainerDeletionEntry> fetchContainerDeletionEntries(
+      List<? extends PartitionId> assignedPartitions) throws CloudStorageException {
+    Set<CosmosContainerDeletionEntry> containerDeletionEntrySet =
+        requestAgent.doWithRetries(() -> cosmosDataAccessor.getDeprecatedContainers(containerDeletionQueryBatchSize),
+            "GetDeprectedContainers", null);
```
Comment (SophieGuo410) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureContainerCompactor.java:260]: minor: GetDeprectedContainers -> GetDeprecatedContainers

**Review by SophieGuo410 — COMMENTED** [review]
Overall lgtm. Added some minor comments.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java [hunk 14]
```diff
@@ -70,10 +70,13 @@
   private static final String START_TIME_PARAM = "@startTime";
   private static final String END_TIME_PARAM = "@endTime";
   private static final String LIMIT_PARAM = "@limit";
+  private static final String ACCOUNT_ID_PARAM = "@accountId";
+  private static final String CONTAINER_ID_PARAM = "@containerId";
   private static final String EXPIRED_BLOBS_QUERY = constructDeadBlobsQuery(CloudBlobMetadata.FIELD_EXPIRATION_TIME);
   private static final String DELETED_BLOBS_QUERY = constructDeadBlobsQuery(CloudBlobMetadata.FIELD_DELETION_TIME);
   private static final String CONTAINER_BLOBS_QUERY =
-      "SELECT TOP %d * FROM c WHERE c.accountId=%d and c.containerId=%d";
+      "SELECT TOP %d " + LIMIT_PARAM + " FROM c WHERE c.accountId=" + ACCOUNT_ID_PARAM + " and c.containerId="
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:78]: I think Cosmos will reject this.  "%d" needs to be removed, and you need the "*" before FROM.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java [hunk 15]
```diff
@@ -70,10 +70,13 @@
...
+          + CONTAINER_ID_PARAM;
   private static final String BULK_DELETE_QUERY = "SELECT c._self FROM c WHERE c.id IN (%s)";
   private static final String DEPRECATED_CONTAINERS_QUERY =
       "SELECT TOP %d * from c WHERE c.deleted=false order by c.deleteTriggerTimestamp";
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:82]: Can you change this to LIMIT_PARAM too?  I missed it earlier.

**Review by lightningrob — CHANGES_REQUESTED** [review]
Fix the Cosmos queries.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java [hunk 16]
```diff
@@ -72,15 +72,15 @@
   private static final String LIMIT_PARAM = "@limit";
   private static final String ACCOUNT_ID_PARAM = "@accountId";
   private static final String CONTAINER_ID_PARAM = "@containerId";
+  private static final String MAX_ENTRIES_PARAM = "@maxEntries";
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:75]: What's the difference between LIMIT and MAX_ENTRIES params?

Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:75]: Removed MAX_ENTRIES.

**Review by ankagrawal — COMMENTED** [review]
Addressed all review comments.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java (hunk 16 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java:75]: It doesn't look like it.

**Review by lightningrob — APPROVED** [review]
LGTM, thanks for making the changes.

**Review by SophieGuo410 — APPROVED** [review]
LGTM!
